Disables the Button during onPress call in PressableWithFeedback

priyeshshah11 opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->

The main purpose of this PR is to disable the button to prevent multiple clicks during onPress call execution.
It updates the `PressableWithFeedback` component to achieve this.  It also updates the `Button` component to replace the usage of `Pressable` with `PressableWithFeedback` component to make use of this for all buttons. It also updates the `createWorkspace` action to return a Promise.

There are some minor changes to `BaseGenericPressable` component as well to fix some errors discovered during implementation.

### Fixed Issues
<!---
1. Please replace GH_LINK with a URL link to the GitHub issue this Pull Request is fixing.
2. Please replace PROPOSAL: GH_LINK_ISSUE(COMMENT) with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<number-of-the-issue>
$ https://github.com/Expensify/App/issues/<number-of-the-issue(comment)>

Do NOT only link the issue number like this: $ #<number-of-the-issue>
--->
$ https://github.com/Expensify/App/issues/14572
PROPOSAL: https://github.com/Expensify/App/issues/14572#issuecomment-1407328690


### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

1. Go to Settings -> Workspaces.
2. Click on New Workspace multiple times quickly
3. Verify only one workspace is created.
- [x] Verify that no errors appear in the JS console

### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->
Same as above

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

1. Go to Settings -> Workspaces.
2. Click on New Workspace multiple times quickly
3. Verify only one workspace is created.
- [x] Verify that no errors appear in the JS console

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://user-images.githubusercontent.com/38547776/235039259-9b4f4bf2-95b9-403f-b586-0a9c0d30816c.mov



<!-- add screenshots or videos here -->

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://user-images.githubusercontent.com/38547776/235039331-e24ec241-06db-4ac5-92ea-0b54b4e6dfe1.mov


<!-- add screenshots or videos here -->

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://user-images.githubusercontent.com/38547776/235039381-56f06a71-7228-470f-a87f-7d786a15e7ff.mov

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Desktop</summary>

https://user-images.githubusercontent.com/38547776/235039469-61a5ae91-237e-4282-9f0d-15486110c29e.mov


<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS</summary>

<!-- add screenshots or videos here -->

https://user-images.githubusercontent.com/38547776/235348901-ff796077-a9c8-4dc2-8564-2441dccea0a4.mov


</details>

<details>
<summary>Android</summary>

https://user-images.githubusercontent.com/38547776/235039526-b0ab0049-207b-4209-890d-62759c8be387.mov


<!-- add screenshots or videos here -->

</details>

priyeshshah11 committed:
added disabling of Button during onPress call in PressableWithFeedback

priyeshshah11 committed:
Merge branch 'main' of https://github.com/priyeshshah11/App into priyeshshah11-disable-buttons

s77rt commented:
@priyeshshah11 Thank you. I will review asap, in the meantime can you please fill the details section of the PR highlighting the main changes here? (e.g. refactor Button to use PressableWithFeedback, etc.)

priyeshshah11 commented:
@s77rt sorry had totally missed that, updated now!

s77rt commented on src/components/Button/index.js:
```diff
@@ -109,6 +109,9 @@ const propTypes = {
     nativeID: PropTypes.string,
+
+    /**  accessibility label for the component */
+    accessibilityLabel: PropTypes.string,
```
This should be required

priyeshshah11 replied:
does that mean all Buttons will need an accessibility label? & should adding all of them be in scope of this PR?

s77rt replied:
Well since `accessibilityLabel` is required on `GenericPressable` it would make sense to have this required on `Button` too. I think it would be better to split this to two PRs:
1. Migrate from Pressable to PressableWithFeedback for Button
2. Disable PressableWithFeedback on press

cc @roryabraham thoughts on this?

roryabraham replied:
I agree that `accessibilityLabel` should be required. I also agree that we should start by disabling `PressableWithFeedback` on press in this PR, then create a separate PR to migrate from `Pressable` to `PressableWithFeedback` for `Button`.

priyeshshah11 replied:
@roryabraham I have already done the migration part in this PR as we didn't get a reply for ~2 weeks, so what's the plan of action here?

s77rt commented on src/components/Button/index.js:
```diff
@@ -254,47 +258,41 @@ class Button extends Component {
-                style={[
+                wrapperStyle={[
                     this.props.isDisabled ? {...styles.cursorDisabled, ...styles.noSelect} : {},
+                    (this.props.isDisabled && (this.props.success || this.props.danger)) ? styles.buttonOpacityDisabled : undefined,
+                    (this.props.isDisabled && !this.props.danger && !this.props.success) ? styles.buttonDisabled : undefined,
```
Why we are moving those from inner style to wrapper style?

priyeshshah11 replied:
that's how the new component works, it'll make sense when you try it out.

s77rt replied:
Can you please elaborate? Those styles were being passed to `<OpacityView />` why we are passing them to `<Pressable />`

s77rt commented on src/components/Pressable/GenericPressable/BaseGenericPressable.js:
```diff
@@ -122,14 +122,14 @@ const GenericPressable = forwardRef((props, ref) => {
             style={state => [
                 getCursorStyle(isDisabled, [props.accessibilityRole, props.role].includes('text')),
-                props.style,
+                StyleUtils.parseStyleFromFunction(props.style, state),
```
Why this change is required?

priyeshshah11 replied:
It is required to apply styles based on the component states like isHovered, isDisabled, etc just like the lines below it. @robertKozik told me to add it so maybe he can add more context here if needed

robertKozik replied:
The idea inside `BaseGenericPressable` was to spread styling based on Pressable state into different props (HoverStyle, focusStyle etc.). So in place of:
```javaScript
style={({isHovered, isPressed}) => ({
    isHovered && <<hoverStyle>>,
    isPressed && <<pressStyle>>
})}
```
We would get:
```javaScript
pressedStyle={<<pressStyle>>}
hoverStyle={<<hoverStyle>>}
```

But as `PressableWithFeedback` shifts all these state-aware styling into `opacityView` and passes `wrapperStyle` as the style prop to `GenericPressable`, I suggested this change in order to still have the possibility of state-aware styling inside the wrapperStyle prop.

During our convo with @priyeshshah11 I came up with this change because I thought It could be used there. 
All in all, even when there is no use here, this change is beneficial.

s77rt commented on src/pages/workspace/WorkspacesListPage.js:
```diff
@@ -200,9 +200,10 @@ class WorkspacesListPage extends Component {
+                        accessibilityLabel={this.props.translate('workspace.new.newWorkspace')}
                         success
                         text={this.props.translate('workspace.new.newWorkspace')}
-                        onPress={() => Policy.createWorkspace()}
+                        onPress={Policy.createWorkspace}
```
Revert this line. This intended because the `onPress` callback will call the `createWorkspace` with a parameter event which will translate to `ownerEmail`.

onPress currently does not pass any parameter (not sure if this is intended though) so that may change.

priyeshshah11 replied:
but didn't @roryabraham say so [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1682377415283179?thread_ts=1682369641.777619&cid=C01GTK53T8Q) ?

priyeshshah11 replied:
I didn't get what you mean here? what has `ownerEmail` got to do here? I might be missing some context

s77rt replied:
`onPress={Policy.createWorkspace}`
This means whatever arguments are passed to `onPress` pass them to `createWorkspace`.

This is similar to:
`onPress={(e) => Policy.createWorkspace(e)}`

`e` here will be the `ownerEmail` which is what the `createWorkspace` accepts as first argument.

s77rt commented on src/libs/actions/Policy.js:
```diff
@@ -949,7 +950,7 @@ function createWorkspace(ownerEmail = '', makeMeAdmin = false, policyName = '',
     });
 
-    Navigation.isNavigationReady()
+    return Navigation.isNavigationReady()
```
Actions should not return a promise. Just revert this. The InteractionManager seems to be enough for our case, so no promise is actually needed.

priyeshshah11 replied:
but didn't @roryabraham say so [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1682377415283179?thread_ts=1682369641.777619&cid=C01GTK53T8Q) ?

s77rt replied:
This may create a pattern for further devs which I'm not sure if we want that.

But do we actually need a promise here?

priyeshshah11 replied:
yes we do need it, as InteractionManager is not enough. And I think we were aware of this pattern change & accepted it before assigning/commencing work on this PR as per @roryabraham's comments in Slack.

s77rt replied:
The promise used here does not wait for onyx. It waits for navigation, it just happen that at that point onyx data is updated.

s77rt replied:
Actually at this point (_the following statement may change_) we are waiting for onyx data update since the optimistic data use onyx `SET` method which blocking (sync).

roryabraham replied:
> The promise used here does not wait for onyx. It waits for navigation, it just happen that at that point onyx data is updated.

Yes, that's a very good point. I think this is maybe fine for now, but long-term we do need a better way to wait for just the optimistic data to be written to cache and for subscribers to be updated.

> optimistic data use onyx SET method which is blocking (sync).

`Onyx.set` is async:

- https://github.com/Expensify/react-native-onyx/blob/0fbe0a4d077f96c4bc30b0b14359f634b271b03b/lib/Onyx.js#L862
- https://github.com/Expensify/react-native-onyx/blob/0fbe0a4d077f96c4bc30b0b14359f634b271b03b/lib/Onyx.js#L778

We could feasibly make a sync Onyx API on iOS and Android right now using JSI, but have no way to do so on web currently. We make `Onyx.set` async to make it more consistent with `Onyx.merge`

priyeshshah11 replied:
@ntdiary If this is not a promise then we don't have a simple way of knowing when this action completed & when to re-enable the button.

priyeshshah11 committed:
code cleanup

s77rt commented:
@priyeshshah11 Can you please fix the hover style (and press style). The button looks too much dimmed than it should

`main`:
![Screenshot from 2023-04-29 01-36-25](https://user-images.githubusercontent.com/16493223/235273451-a64bfaa9-2e7b-45ed-994e-a596d230c0ff.png)


PR:
![Screenshot from 2023-04-29 01-37-13](https://user-images.githubusercontent.com/16493223/235273482-514d4577-fea4-4512-9c1d-59988f4c42eb.png)

priyeshshah11 commented:
> @priyeshshah11 Can you please fix the hover style (and press style). The button looks too much dimmed than it should
> 

So earlier we were using the same dimming value for hover & press which was 0.5 & now `PressableWithFeedback` correctly uses the `pressDimValue` (0.2) when pressed & `pressHoverValue` (0.5) when hovered. Let me know what we want.

cc: Hi 👋 @shawnborton 😄

priyeshshah11 committed:
Merge branch 'main' of https://github.com/priyeshshah11/App into priyeshshah11-disable-buttons

s77rt commented:
It looks too dim for a hover state :sweat_smile: Let's get this to reflect the staging values

s77rt commented on src/components/Button/index.js:
```diff
@@ -254,47 +258,41 @@ class Button extends Component {
+                    this.props.large ? styles.buttonLarge : undefined,
+                    this.props.success ? styles.buttonSuccess : undefined,
+                    this.props.danger ? styles.buttonDanger : undefined,
+                    (this.props.success && isHovered && !isDisabled) ? styles.buttonSuccessHovered : undefined,
+                    (this.props.danger && isHovered && !isDisabled) ? styles.buttonDangerHovered : undefined,
```
The style here is not correct. Previously we had `const activeAndHovered = !this.props.isDisabled && hovered;` where the active state was solely base on the `isDisabled` prop. Now it's based on the `isDisabled` state which is based on both `isDisabled` prop and `isLoading` prop.

priyeshshah11 replied:
you're right for both the above comments, changes applied.

s77rt commented on src/components/Pressable/PressableWithFeedback.js:
```diff
@@ -24,9 +25,34 @@ const PressableWithFeedbackDefaultProps = {
+            {...propsWithoutStyling}
+            disabled={disabled}
+            onPress={(e) => {
+                if (disabled) { return; }
```
Is this required? If the button is disabled how would the `onPress` even get called?

priyeshshah11 replied:
you're correct, I had added it as for some reason I was still able to click it while initial testing but couldn't reproduce it now so have removed it.

priyeshshah11 committed:
code cleanup

priyeshshah11 commented:
> It looks too dim for a hover state 😅 Let's get this to reflect the staging values

happy to change it to whatever needed, but can we please get a confirmation from @robertKozik & @shawnborton first? As I believe @robertKozik added that value [here](https://github.com/Expensify/App/commit/ee2e3b08c47956252effa35467035d23b9a4a2f8#diff-4c815c48cfd171056c4be0e2bc556302e115c20c2bef5b5b00dbe55450faca5a)

shawnborton commented:
Hmm we already have a value for the hover color right? And then when it is actively being pressed, we can use the .2 opacity dim.

priyeshshah11 commented:
> Hmm we already have a value for the hover color right? And then when it is actively being pressed, we can use the .2 opacity dim.

yup that's correct, hover has greenHover colour & a 0.5 opacity dim & press has 0.2 opacity dim.

priyeshshah11 commented:
@robertKozik could you please share your thoughts on why you had added/used `hoverDimmingValue` on hover? & is it ok for me to remove it if the design suggests so?

priyeshshah11 committed:
fixed hover & pressed styles

priyeshshah11 commented:
@s77rt I have made the changes as per Shawn's comments above, i.e. use hover colours when hovered (without any dim) & use 0.2 opacity dim when pressed.

s77rt commented on src/components/Button/index.js:
```diff
@@ -254,47 +258,43 @@ class Button extends Component {
-                    );
-                }}
-            </Pressable>
+                <View>
+                    {this.renderContent()}
+                    {this.props.isLoading && (
+                        <ActivityIndicator
+                            color={(this.props.success || this.props.danger) ? themeColors.textLight : themeColors.text}
+                            style={[styles.pAbsolute, styles.l0, styles.r0]}
+                        />
+                    )}
+                </View>
```
```suggestion
                <>
                    {this.renderContent()}
                    {this.props.isLoading && (
                        <ActivityIndicator
                            color={(this.props.success || this.props.danger) ? themeColors.textLight : themeColors.text}
                            style={[styles.pAbsolute, styles.l0, styles.r0]}
                        />
                    )}
                </>
```

s77rt reviewed (changes requested):
Just a slight change

s77rt commented:
@priyeshshah11 Thank you

@shawnborton Isn't 0.2 too dim compared to staging or this is actually intended?
![Screenshot from 2023-05-01 14-57-27](https://user-images.githubusercontent.com/16493223/235465364-ede00608-ea3c-4e88-b564-9634f98e1b83.png)
![Screenshot from 2023-05-01 14-57-21](https://user-images.githubusercontent.com/16493223/235465368-f6f12378-d883-4069-8784-a6f36a8e31f7.png)

priyeshshah11 committed:
replace View with frgament

shawnborton commented:
Hmm we should just be using what we currently use, there's no need to change the style of the pressable opacity. What do we currently do on staging/production, and why is it different here?

shawnborton commented:
Got it, apologies as I am having a hard time following along here. 

At what point was that change introduced? Can you link to the PR/issue where we changed it to .2? .2 is probably fine but just want to make sure I have all of the context.

priyeshshah11 commented:
@shawnborton here is the PR https://github.com/Expensify/App/pull/17404/

shawnborton commented:
Cool, let's follow that then and just use the same value that PR is using. So that means that on press, we dim the button by .2, leaving a .8 opaque button right?

priyeshshah11 commented:
@shawnborton It actually sets the opacity to 0.2 (the second image [here](https://github.com/Expensify/App/pull/18122#issuecomment-1529763771))

shawnborton commented:
Yeah, that doesn't seem right. Can we just use what we are using on staging? And can you confirm what that value is?

priyeshshah11 commented:
> Yeah, that doesn't seem right. Can we just use what we are using on staging? And can you confirm what that value is?

It is 0.5

priyeshshah11 committed:
changed the pressDimValue back to 0.5 as advised by the design team

priyeshshah11 commented:
@s77rt updated it to 0.5 (which is the same as staging/prod) as requested. 

@shawnborton I don't where the 0.2 came from, maybe @robertKozik might be able to explain it better.

s77rt commented:
@priyeshshah11 We have a bug (and some failing tests?)

1. Go to Settings -> Payments
2. Click the Add payment method button
3. Fix console errors and verify the modal popover shows up

priyeshshah11 committed:
passing the click events down from BaseGenericPressable to fix a bug

priyeshshah11 commented:
@s77rt Fixed the bug, but I don't think the failing tests are related to this PR, are they?

s77rt commented:
## Reviewer Checklist

- [X] I have verified the author checklist is complete (all boxes are checked off).
- [X] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [X] I verified the steps for local testing are in the `Tests` section
    - [X] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [X] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [X] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [X] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [X] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [X] Android / native
    - [X] Android / Chrome
    - [X] iOS / native
    - [X] iOS / Safari
    - [X] MacOS / Chrome / Safari
    - [X] MacOS / Desktop
- [X] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [X] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [X] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [X] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [X] I verified that comments were added to code that is not self explanatory
    - [X] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [X] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [X] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [X] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [X] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [X] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [X] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [X] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [X] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [X] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [X] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [X] If a new component is created I verified that:
    - [X] A similar component doesn't exist in the codebase
    - [X] All props are defined accurately and each prop has a `/** comment above it */`
    - [X] The file is named correctly
    - [X] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [X] The only data being stored in the state is data necessary for rendering and nothing else
    - [X] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [X] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [X] All JSX used for rendering exists in the render method
    - [X] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [X] If any new file was added I verified that:
    - [X] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [X] If a new CSS style is added I verified that:
    - [X] A similar style doesn't already exist
    - [X] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [X] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [X] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [X] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [X] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [X] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [X] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

<!-- add screenshots or videos here -->

https://user-images.githubusercontent.com/16493223/235504523-0dd21299-a909-40af-84eb-9d43ca95c918.mp4


</details>

<details>
<summary>Mobile Web - Chrome</summary>

<!-- add screenshots or videos here -->

https://user-images.githubusercontent.com/16493223/235504497-131768b7-2143-459f-8822-0d594d251078.mp4


</details>

<details>
<summary>Mobile Web - Safari</summary>

<!-- add screenshots or videos here -->

https://user-images.githubusercontent.com/16493223/235504480-0704b613-2b5f-46e2-84c5-348857cb0646.mp4


</details>

<details>
<summary>Desktop</summary>

<!-- add screenshots or videos here -->

https://user-images.githubusercontent.com/16493223/235504331-302fdd6d-7ef6-4b6d-ba84-9a81c4bbb3bd.mp4


</details>

<details>
<summary>iOS</summary>

<!-- add screenshots or videos here -->

https://user-images.githubusercontent.com/16493223/235504292-b334f731-84be-41ce-99f5-bd40a09557da.mp4


</details>

<details>
<summary>Android</summary>

<!-- add screenshots or videos here -->

https://user-images.githubusercontent.com/16493223/235504280-5d6007e9-316c-4d9f-ae1c-0c287296dee3.mp4


</details>

s77rt commented:
@priyeshshah11 Please update test steps, it's go to Settings not Profile.

There is another bug that I don't think is originating from this PR but still worth mentioning, disabled buttons are pressable.


https://user-images.githubusercontent.com/16493223/235506995-90df0e31-686a-4806-acea-4543edffd709.mp4

priyeshshah11 committed:
restricts press events when the button is disabled

priyeshshah11 commented:
> @priyeshshah11 Please update test steps, it's go to Settings not Profile.
> 
> There is another bug that I don't think is originating from this PR but still worth mentioning, disabled buttons are pressable.
> 

Updated the tests & also fixed that bug by passing `disabled={isDisabled}` in the `BaseGenericPressable`.

s77rt commented:
@priyeshshah11 The last commit made the cursor no longer customizable, we used to have `not-allowed` cursor on disabled buttons.

@robertKozik A little help on this one, why disabled prop is not being passed to Pressable, is it intended for cursor customisation? And if that was the only case, should we use `pointer-events` to work on top of that?

robertKozik commented:
Hi all @s77rt @priyeshshah11 ! sorry for keep you waiting, I was out of work for some time. 
1. The dim value for both press and hover is originated from the [Design Doc](https://docs.google.com/document/d/1yHHs4IfaMxB2q7eW8Hyt7ew4rNtY6HdduPKdf-Q2ZxE/edit?pli=1#bookmark=id.z63gelgjqb9o) for the in-house Pressable implementation. `0.2` is in line with default value of `TouchableOpacity` `activeOpacity` prop value. Hover dim value originated from the default value of the `OpacityView`. Both values are the default values only though, so it can be freely changed based on the use-case.
2. @s77rt you're right about the customisation. Disabled state of the component wasn't passed to the underlying react-native Pressable, as in `react-native-web` disabled prop: _"Disables all pointer interactions with the element."_ and we want to preserve the possibility to change the cursor appearance for disabled components. 
The bug you encountered @s77rt I think is more connected with the condition of dimming with `PressableWithFeedback` as it's not taking under consideration whether Pressable is disabled or not. onPress actions won't be fired as we are checking the disabled state inside the function wrapper instead

Also @priyeshshah11 is there any particular reason to implement the delay logic inside the `PressableWithFeedback` rather than in the `BaseGenericPressable` ?

priyeshshah11 committed:
removed passing down the disabled prop in BaseGenericPressable & fixed dimming effect on press when the button is disabled.

priyeshshah11 commented:
> Also @priyeshshah11 is there any particular reason to implement the delay logic inside the `PressableWithFeedback` rather than in the `BaseGenericPressable` ?

Not really, mainly because @roryabraham suggested so [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1682442972139729?thread_ts=1682369641.777619&cid=C01GTK53T8Q)

@s77rt Fixed the other issues.

> Got it, so sounds like we would dim by .2 and thus the element remains at .8? I like the idea of standardizing on .8 opacity for the pressed state everywhere, it feels more subtle. Does that sound good?

@shawnborton as per the design doc & @robertKozik's comment above, I think it's the opposite of what you're saying here.

priyeshshah11 commented on src/components/Pressable/PressableWithFeedback.js:
```diff
@@ -12,28 +13,50 @@ const omittedProps = ['style', 'pressStyle', 'hoverStyle', 'focusStyle', 'wrappe
                     style={[
                         StyleUtils.parseStyleFromFunction(props.style, state),
-                        state.pressed && StyleUtils.parseStyleFromFunction(props.pressStyle, state),
+                        !disabled && state.pressed && StyleUtils.parseStyleFromFunction(props.pressStyle, state),
                         state.hovered && StyleUtils.parseStyleAsArray(props.hoverStyle, state),
                         state.focused && StyleUtils.parseStyleAsArray(props.focusStyle, state),
```
@s77rt @robertKozik do you guys think I should add the `!disabled` check for the below styles too? @shawnborton that would mean no hover or focus effect when the button is disabled which I think makes sense but just need confirmation from you guys

shawnborton commented:
@robertKozik let me know where the best place is to discuss this, but I think having elements at .2 opacity onPress is just way too dim. Can we change that to .8 please?

robertKozik commented on src/components/Pressable/PressableWithFeedback.js:
```diff
@@ -12,28 +13,50 @@ const omittedProps = ['style', 'pressStyle', 'hoverStyle', 'focusStyle', 'wrappe
-        // eslint-disable-next-line react/jsx-props-no-spreading
-        <GenericPressable ref={ref} style={props.wrapperStyle} {...propsWithoutStyling}>
+        <GenericPressable
+            ref={ref}
+            style={props.wrapperStyle}
+            // eslint-disable-next-line react/jsx-props-no-spreading
+            {...propsWithoutStyling}
+            disabled={disabled}
+            onPress={(e) => {
+                setDisabled(true);
+                const onPress = props.onPress(e);
+                InteractionManager.runAfterInteractions(() => {
+                    if (!(onPress instanceof Promise)) {
+                        setDisabled(props.disabled);
+                        return;
+                    }
+                    onPress.then(() => {
+                        setDisabled(props.disabled);
+                    });
+                });
+            }}
+        >
             {state => (
                 <OpacityView
-                    shouldDim={state.pressed || state.hovered}
-                    dimmingValue={state.pressed ? props.pressDimmingValue : props.hoverDimmingValue}
+                    shouldDim={!disabled && state.pressed}
+                    dimmingValue={props.pressDimmingValue}
```
I wouldn't strip this from the functionality of distinguish dimming value between hover and press. As now they are unified on the `variables.js` level I think old logic can stay

priyeshshah11 replied:
however @shawnborton mentioned that we do not want any dimming on hover, there is only a colour change when hovered, which is applied in `Button` styles

robertKozik replied:
In button case it's true, but Pressable component can be used in different places also, where it can come in handy IMO

priyeshshah11 replied:
could you please give an example? or can we all discuss this in Slack & get a conclusion?

robertKozik commented:
> @robertKozik let me know where the best place is to discuss this, but I think having elements at .2 opacity onPress is just way too dim. Can we change that to .8 please?

Sure, I don't see any problem. I'm not sure where is the place to discuss this. Maybe @roryabraham will know, as he was supervising the `Accessible Pressable` Design Doc on the Expensify side

priyeshshah11 committed:
cleanup related to hover & pressed styles

priyeshshah11 commented:
@s77rt @roryabraham I have addressed all the above issues after discussions with @robertKozik & @shawnborton. Hopefully, that's all to get this merged in. I'll be OOO till tomorrow.

s77rt commented:
@shawnborton @robertKozik Thank you!

@priyeshshah11 Can you please check the failing test, I have checked some random PRs and none has that failing test, so it's probably related to a change here.

priyeshshah11 commented:
> @priyeshshah11 Can you please check the failing test, I have checked some random PRs and none has that failing test, so it's probably related to a change here.

@s77rt As mentioned before, these changes have nothing to do with that test failing. The same thing was happening in this PR https://github.com/Expensify/App/pull/15130, please refer to this comment from @mountiny https://github.com/Expensify/App/pull/15130#pullrequestreview-1372740288

priyeshshah11 committed:
fixes unread indicator test

priyeshshah11 commented:
I think I have fixed it locally, let's see if that fixes it.

priyeshshah11 commented:
@s77rt all tests are passing now!

s77rt reviewed (dismissed):
LGTM! :rocket: 

cc @roryabraham

mananjadhav commented on src/components/Button/index.js:
```diff
@@ -109,6 +109,9 @@ const propTypes = {
     /** Id to use for this button */
     nativeID: PropTypes.string,
+
+    /**  accessibility label for the component */
```
```suggestion
    /** Accessibility label for the component */
```

mananjadhav commented on src/libs/Accessibility/index.js:
```diff
@@ -8,7 +8,7 @@ const useScreenReaderStatus = () => {
         const subscription = AccessibilityInfo.addEventListener('screenReaderChanged', setIsScreenReaderEnabled);
 
-        return subscription.remove;
+        return subscription && subscription.remove;
```
why do we need this return?

priyeshshah11 replied:
to fix broken tests

mananjadhav commented on src/components/Pressable/PressableWithFeedback.js:
```diff
@@ -24,17 +25,41 @@ const PressableWithFeedbackDefaultProps = {
+                setDisabled(true);
+                const onPress = props.onPress(e);
+                InteractionManager.runAfterInteractions(() => {
+                    if (!(onPress instanceof Promise)) {
```
nab, but if onPress is not an instance of Promise, does it still have to be inside `InteractionManager`?

could you elaborate difference between the above code and the following:

```
const onPress = props.onPress(e);
if (!(onPress instanceof Promise)) {
   setDisabled(props.disabled);
   return;
}

InteractionManager.runAfterInteractions(() =>
  onPress.then(() => {
                          setDisabled(props.disabled);
                      });
});
```

mananjadhav replied:
> on some android devices, there is still a possibility for it to capture multiple interactions before onPress's completion

I am not sure about this when we have sync calls, but it's okay to keep what it is right now.

mananjadhav commented on src/styles/variables.js:
```diff
@@ -133,5 +133,5 @@ export default {
     googleEmptyListViewHeight: 14,
     hoverDimValue: 0.5,
-    pressDimValue: 0.2,
+    pressDimValue: 0.8,
```
pardon my ignorance, this is needed because?

priyeshshah11 replied:
that's what the design team wants, feel free to checkout the PR discussion & comments to get more context.

mananjadhav replied:
Got it sorry about that, I was still going through the comments. The tests, or proposal doesn't cover this change. Hence my confusion.

mananjadhav commented on src/components/Button/index.js:
```diff
@@ -254,47 +258,44 @@ class Button extends Component {
+                ]}
                 nativeID={this.props.nativeID}
+                accessibilityLabel={this.props.accessibilityLabel}
+                hoverDimmingValue={1}
```
for this one too, could you help why did we need to set it as 1?

priyeshshah11 replied:
same comment as above

mananjadhav reviewed (commented):
@priyeshshah11 I have small minor comments/questions. Could you take a look and then I can test at my end too?

priyeshshah11 committed:
Update src/components/Button/index.js

Co-authored-by: Manan <manan.jadhav@gmail.com>

priyeshshah11 commented:
@roryabraham & @slafortune would it be fair to ask to increase the job price based on the increase in scope & effort than initially planned?

mananjadhav commented:
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://user-images.githubusercontent.com/3069065/236008870-2d542c55-cb76-4e30-9503-6aa3ebed001c.mov

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://user-images.githubusercontent.com/3069065/236009171-1c443d82-a2c1-4455-9992-eb5d65d3f9d3.mov

</details>

<details>
<summary>Mobile Web - Safari</summary>

</details>

<details>
<summary>Desktop</summary>

https://user-images.githubusercontent.com/3069065/236009232-8f092bce-d167-4f56-871a-1914cc3e3204.mov

</details>

<details>
<summary>iOS</summary>

https://user-images.githubusercontent.com/3069065/236009273-13036843-8d9a-42b9-b4d5-698cdc665544.mov

</details>

<details>
<summary>Android</summary>

https://user-images.githubusercontent.com/3069065/236009318-be4dadf5-8ef0-4ff8-9948-50d73bc40771.mov

</details>

hungvu193 commented:
Hello guys, I'm working on an issue that need to use PressableWithFeedback and I realized when user first try to click the Submit button, the cursor will briefly change into none-select cursor for a while. Is this a feature?

https://user-images.githubusercontent.com/16502320/236152132-7a1f23a1-5b15-4332-a161-340ae0675c20.mov

priyeshshah11 commented:
> Hello guys, I'm working on an issue that need to use PressableWithFeedback and I realized when user first try to click the Submit button, the cursor will briefly change into none-select cursor for a while. Is this a feature?
> 
> 
> 
> https://user-images.githubusercontent.com/16502320/236152132-7a1f23a1-5b15-4332-a161-340ae0675c20.mov
> 
> 
> 
> 

That happens when a button is disabled, and we disable the button when initially clicked until the onPress is completed to avoid multiple quick clicks. But that should be only in this PR, are you using this PR in your case? could you link that issue for reference?

hungvu193 commented:
> That happens when a button is disabled, and we disable the button when initially clicked until the onPress is completed to avoid multiple quick clicks. But that should be only in this PR, are you using this PR in your case? could you link that issue for reference?

@priyeshshah11  I'm working on this issue (https://github.com/Expensify/App/issues/17103), I need to replace TouchableOpacity with Pressable but still want to keep the activeOpacity behavior, so I decided to use PressableWithFeedback, it's working fine with mWeb, but with Web version, it will look like this:

https://user-images.githubusercontent.com/16502320/236154830-347c43be-a417-429e-bd5a-0344f7482c01.mov

So I have questions:
- Should we add more props to PressableWithFeedback that allow user to interact with button multiple times in very short moment (similar with takeEvery and takeLatest in redux-saga).
- In case this is feature, is there any way to prevent that or is there any alternative solution?
Thank you.

s77rt commented:
@priyeshshah11 @roryabraham Can you please address @hungvu193's concern. Should we add an option to opt out from the "disable button on click" functionality?

priyeshshah11 commented:
@hungvu193 I cannot answer the 1st point as that's for @roryabraham to decide but I don't see any harm in adding it if you need it. For the second point, currently there is no way to prevent this.

Either way, it doesn't need to hold up this PR.

cc: @s77rt

roryabraham commented:
> @robertKozik let me know where the best place is to discuss this, but I think having elements at .2 opacity onPress is just way too dim. Can we change that to .8 please?

If @shawnborton thinks this is best, I think we can make that change without further discussion

roryabraham commented:
> @roryabraham & @slafortune would it be fair to ask to increase the job price based on the increase in scope & effort than initially planned?

@priyeshshah11 the issue is the better place to ask this question than in the PR

roryabraham reviewed (changes requested):
Interesting concern from @hungvu193, and not one that I considered before. I think it _is_ a feature that:

- When a Pressable becomes disabled, then we show the none-select cursor.
- When a Pressable is pressed, it becomes disabled until `onPress` completes to prevent multiple repeated interactions.

However, when these things both happen in a very short time period, it looks janky, and that's a problem we should solve.

In order to solve that problem, I suggest a naïve workaround to only change the cursor style after a brief timeout. Something like this (haven't tested but this should illustrate the idea):

```js
const isDisabled = useMemo(() => {
    let shouldBeDisabledByScreenReader = false;
    if (enableInScreenReaderStates === CONST.SCREEN_READER_STATES.ACTIVE) {
        shouldBeDisabledByScreenReader = !isScreenReaderActive;
    }
    
    if (enableInScreenReaderStates === CONST.SCREEN_READER_STATES.DISABLED) {
        shouldBeDisabledByScreenReader = isScreenReaderActive;
    }
    
    return props.disabled || shouldBeDisabledByScreenReader;
}, [isScreenReaderActive, enableInScreenReaderStates, props.disabled]);
  
const [shouldUseDisabledCursor, setShouldUseDisabledCursor] = useState(isDisabled);
useEffect(() => {
    if (!isDisabled) {
        setShouldUseDisabledCursor(isDisabled);
    } else {
        setTimeout(() => setShouldUseDisabledCursor(isDisabled), 1000);
    }
}, [isDisabled]);


...
...

getCursorStyle(shouldUseDisabledCursor, [props.accessibilityRole, props.role].includes('text')),
```

priyeshshah11 commented:
> Interesting concern from @hungvu193, and not one that I considered before. I think it _is_ a feature that:

@roryabraham I assume you're asking for this feature to be implemented in a seperate PR as it wasn't in the scope for this PR, right?

priyeshshah11 committed:
Merge branch 'main' of https://github.com/priyeshshah11/App into priyeshshah11-disable-buttons

priyeshshah11 commented:
> > Interesting concern from @hungvu193, and not one that I considered before. I think it _is_ a feature that:
> 
> @roryabraham I assume you're asking for this feature to be implemented in a seperate PR as it wasn't in the scope for this PR, right?

@roryabraham could you please confirm? or merge this PR if it's good to go?

cc: @s77rt @mananjadhav

s77rt commented:
I'm okay going either way. Deferring to @roryabraham

roryabraham reviewed (approved):
Ok, we can handle this in a follow-up

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/4997415254) to production by https://github.com/yuwenmemon in version: 1.3.14-14 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

s77rt commented on src/libs/actions/Policy.js:
```diff
@@ -1078,7 +1079,7 @@ function createWorkspace(ownerEmail = '', makeMeAdmin = false, policyName = '',
     );
 
-    Navigation.isNavigationReady().then(() => {
+    return Navigation.isNavigationReady().then(() => {
```
Oops we were actually supposed to return a value on `Button`. We still need the promise though.

@priyeshshah11 Can you please raise a quick follow up PR?

s77rt replied:
Looks like we don't need it.

priyeshshah11 replied:
@s77rt what's the conclusion here? do you still want me to raise another PR? to remove the promise part or instead return it from the `onPress`? My view is that we should keep it & return it but up to you.

s77rt replied:
@priyeshshah11 Raise a PR that simply return the onPress on Button. (promise may be removed on the other PR)

s77rt commented on src/components/Pressable/PressableWithFeedback.js:
```diff
@@ -24,21 +25,41 @@ const PressableWithFeedbackDefaultProps = {
+            onPress={(e) => {
+                setDisabled(true);
+                const onPress = props.onPress(e);
+                InteractionManager.runAfterInteractions(() => {
+                    if (!(onPress instanceof Promise)) {
+                        setDisabled(props.disabled);
```
The `props.disabled` that we are using inside the `InteractionManager.runAfterInteractions` callback or the one after the promise is resolved (few lines below) is evaluated at the time we call onPress and that value is used later. Meaning props.disabled does not reflect the current state but the state that we evaluated when onPress was initially called.

To better explain, here is an example:
1. props.disabled is false
2. Call onPress
3. Evaluate the promise callback, After the promise is resolved we will call setDisabled(false) // props.disabled is evaluated now
4. Promise is not resolved yet (button action still being executed)
5. props.disabled is set to true
6. Promise is resolved
7. Now we will call the previously evaluated callback which is setDisabled(false) even though the current props.disabled is true


This logic is copied from OptionRow but we missed that case in both PRs which lead to a regression https://github.com/Expensify/App/issues/20983.
